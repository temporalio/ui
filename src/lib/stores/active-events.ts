import { get, writable } from 'svelte/store';

import type { EventGroup } from '$lib/models/event-groups/event-groups';
import type { WorkflowEvent } from '$lib/types/events';

export const indexPageSize = 200;
export const startIndex = writable(0);
export const endIndex = writable(indexPageSize);

export const activeEvents = writable<string[]>([]);
export const activeGroups = writable<string[]>([]);

export const clearActives = () => {
  activeGroups.set([]);
  activeEvents.set([]);
  startIndex.set(0);
  endIndex.set(indexPageSize);
};

export const clearActiveEvents = () => {
  activeGroups.set([]);
  activeEvents.set([]);
};

export const setActiveGroup = (group: EventGroup) => {
  activeEvents.set([]);
  if (!get(activeGroups).includes(group.id)) {
    activeGroups.set([...get(activeGroups), group.id]);
  } else {
    activeGroups.set(get(activeGroups).filter((id) => id !== group.id));
  }
};

export const setSingleActiveGroup = (group: EventGroup) => {
  activeEvents.set([]);
  if (!get(activeGroups).includes(group.id)) {
    activeGroups.set([group.id]);
  } else {
    activeGroups.set([]);
  }
};

export const setSingleActiveEvent = (event: WorkflowEvent) => {
  activeGroups.set([]);
  if (!get(activeEvents).includes(event.id)) {
    activeEvents.set([event.id]);
  } else {
    activeEvents.set([]);
  }
};

export const setActiveEvent = (event: WorkflowEvent, group: EventGroup) => {
  activeGroups.set([]);

  if (!get(activeEvents).includes(event.id)) {
    const activeEventInGroup = get(activeEvents).find(
      (id) => group?.eventIds.has(id),
    );
    if (activeEventInGroup) {
      activeEvents.set([
        ...get(activeEvents).filter((id) => id !== activeEventInGroup),
        event.id,
      ]);
    } else {
      activeEvents.set([...get(activeEvents), event.id]);
    }
  } else {
    activeEvents.set(get(activeEvents).filter((id) => id !== event.id));
  }
};

export const setActiveGroupAndEvent = (
  group: EventGroup,
  event: WorkflowEvent,
) => {
  if (group) {
    if (!get(activeGroups).includes(group.id)) {
      activeGroups.set([...get(activeGroups), group.id]);
    } else {
      activeGroups.set(get(activeGroups).filter((id) => id !== group.id));
    }
  }

  if (event) {
    if (!get(activeEvents).includes(event.id)) {
      activeEvents.set([...get(activeEvents), event.id]);
    } else {
      activeEvents.set(get(activeGroups).filter((id) => id !== event.id));
    }
  }
};
