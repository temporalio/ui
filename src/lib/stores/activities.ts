import { derived, get, Readable } from 'svelte/store';
import {
  isActivityTaskScheduledEvent,
  isActivityTaskStartedEvent,
  isActivityTaskCompletedEvent,
  isActivityTaskFailedEvent,
  isActivityTaskTimedOutEvent,
  isActivityTaskCancelRequestedEvent,
  isActivityTaskCanceledEvent,
} from '$lib/utilities/is-event-type';

type ActivitiesMap = { [key: string]: Activity };
type ActivityStore = Readable<Activity[]> & {
  getActivity: (id: string | number) => Readable<Activity>;
};

const setActivityId = (activities: ActivitiesMap) => (id: string) => {
  if (!activities[id]) activities[id] = {};
  activities[id].id = id;
};

const setActivityStatus =
  (activities: ActivitiesMap) => (id: string, status: ActivityStatus) => {
    if (!activities[id]) activities[id] = {};
    activities[id].status = status;
  };

const addToActivities =
  (activities: ActivitiesMap) =>
  (
    id: string,
    property: keyof Omit<Activity, 'id' | 'status'>,
    event: HistoryEvent,
  ) => {
    if (!activities[id]) activities[id] = {};
    activities[id][property] = event;
  };

export const createActivityStore = <
  E extends HistoryEvent[],
  S extends Readable<E>,
>(
  eventStore: S,
): ActivityStore => {
  const store = derived(
    eventStore,
    (events) => {
      const activities: ActivitiesMap = {};
      const add = addToActivities(activities);
      const setId = setActivityId(activities);
      const setStatus = setActivityStatus(activities);

      for (const event of events) {
        if (isActivityTaskScheduledEvent(event)) {
          const id = event.activityTaskScheduledEventAttributes.activityId;
          add(id, 'activityTaskScheduledEvent', event);
          setId(id);
          setStatus(id, 'Started');
        }

        if (isActivityTaskStartedEvent(event)) {
          const id = String(
            event.activityTaskStartedEventAttributes.scheduledEventId,
          );
          add(id, 'activityTaskStartedEvent', event);
          setStatus(id, 'Started');
        }

        if (isActivityTaskCompletedEvent(event)) {
          const id = String(
            event.activityTaskCompletedEventAttributes.scheduledEventId,
          );
          add(id, 'activityTaskCompletedEvent', event);
          setStatus(id, 'Completed');
        }

        if (isActivityTaskFailedEvent(event)) {
          const id = String(
            event.activityTaskFailedEventAttributes.scheduledEventId,
          );
          add(id, 'activityTaskFailedEvent', event);
          setStatus(id, 'Failed');
        }

        if (isActivityTaskTimedOutEvent(event)) {
          const id = String(
            event.activityTaskTimedOutEventAttributes.scheduledEventId,
          );
          add(id, 'activityTaskTimedOutEvent', event);
          setStatus(id, 'TimedOut');
        }

        if (isActivityTaskCancelRequestedEvent(event)) {
          const id = String(
            event.activityTaskCancelRequestedEventAttributes.scheduledEventId,
          );
          add(id, 'activityTaskCancelRequestedEvent', event);
          setStatus(id, 'CancelRequested');
        }

        if (isActivityTaskCanceledEvent(event)) {
          const id = String(
            event.activityTaskCanceledEventAttributes.scheduledEventId,
          );
          add(id, 'activityTaskCanceledEvent', event);
          setStatus(id, 'Canceled');
        }
      }
      return activities;
    },
    {},
  );

  const activities = derived(store, ($store) => Object.values($store));

  return {
    ...activities,
    getActivity: (id) => derived(store, ($store) => get(store)[id] || {}),
  };
};
