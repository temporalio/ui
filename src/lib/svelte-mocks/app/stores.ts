import { readable } from 'svelte/store';

import type { Settings } from '$lib/types/global';

interface Page<Params extends Record<string, string> = Record<string, string>> {
  url: URL;
  params: Params;
  routeId: string | null;
  data: App.PageData;
  status: number;
  error: Error | null;
}

const settings: Settings = {
  auth: {
    enabled: false,
    options: null,
  },
  baseUrl: 'http://localhost:3000',
  codec: {
    endpoint: '',
    passAccessToken: false,
    includeCredentials: false,
  },
  defaultNamespace: 'default',
  disableWriteActions: false,
  showTemporalSystemNamespace: false,
  batchActionsDisabled: false,
  workflowResetDisabled: false,
  workflowCancelDisabled: false,
  workflowSignalDisabled: false,
  workflowTerminateDisabled: false,
  hideWorkflowQueryErrors: false,
  notifyOnNewVersion: true,
  feedbackURL: '',
  runtimeEnvironment: {
    isCloud: false,
    isLocal: true,
    envOverride: true,
  },
  version: '2.0.0',
};

const data: App.PageData = {
  settings,
};

export const page = readable<Page>({
  error: null,
  params: {
    namespace: 'default',
  },
  routeId: 'namespaces/[namespace]/workflows@root',
  status: 200,
  data,
  url: new URL(
    'http://localhost:3000/namespaces/default/workflows?search=basic&query=WorkflowType%3D%22testing%22',
  ),
});
