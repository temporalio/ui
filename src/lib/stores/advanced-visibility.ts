import { derived } from 'svelte/store';

import { page } from '$app/stores';

import {
  advancedVisibilityEnabled,
  advancedVisibilityEnabledWithOrderBy,
} from '$lib/utilities/advanced-visibility-enabled';

import { cluster } from './cluster';
import { temporalVersion } from './versions';

export const isCloud = derived(
  [page],
  ([$page]) => $page.data?.settings?.runtimeEnvironment?.isCloud,
);

export const supportsAdvancedVisibility = derived(
  [cluster, temporalVersion, isCloud],
  ([$cluster, $temporalVersion, $isCloud]) =>
    advancedVisibilityEnabled($cluster, $temporalVersion) || $isCloud,
);

export const supportsAdvancedVisibilityWithOrderBy = derived(
  [cluster, isCloud],
  ([$cluster, $isCloud]) =>
    advancedVisibilityEnabledWithOrderBy($cluster) || $isCloud,
);
