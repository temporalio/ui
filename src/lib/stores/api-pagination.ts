import type { Readable } from 'svelte/store';
import { derived, writable } from 'svelte/store';

import { page } from '$app/stores';

import {
  defaultItemsPerPage,
  options,
  perPageFromSearchParameter,
  perPageKey,
} from './pagination';

type PaginationMethods<T> = {
  nextPageWithItems: (t: string, items: T[]) => void;
  nextPage: () => void;
  previousPage: () => void;
  setUpdating: () => void;
  reset: () => void;
  resetPageSize: (pageSize: number) => void;
  nextRow: () => void;
  previousRow: () => void;
  setActiveIndex: (activeIndex: number) => void;
};

type PaginationItems<T> = {
  key: string;
  loading: boolean;
  updating: boolean;
  hasNext: boolean;
  hasPrevious: boolean;
  index: number;
  previousPageSize: number;
  pageSize: number;
  indexData: Record<
    number,
    { nextToken: string; start: number; end: number; items: T[] }
  >;
  visibleItems: T[];
  hasNextIndexData: boolean;
  indexStart: number;
  indexEnd: number;
  activeIndex: number;
};

const defaultStore = {
  key: perPageKey,
  loading: true,
  updating: false,
  hasNext: false,
  hasPrevious: false,
  index: 0,
  previousPageSize: defaultItemsPerPage,
  pageSize: defaultItemsPerPage,
  indexData: {},
  visibleItems: [],
  hasNextIndexData: false,
  indexStart: 0,
  indexEnd: 0,
  activeIndex: 0,
};

export type PaginationStore<T> = PaginationMethods<T> &
  Readable<PaginationItems<T>>;

const setFirstOption = (options: string[] | number[]) => {
  const defaultOption = options[0];
  if (!defaultOption) return defaultItemsPerPage;
  const optionAsInt = parseInt(defaultOption.toString());
  if (isNaN(optionAsInt)) return defaultItemsPerPage;
  return optionAsInt;
};

export const getInitialPageSize = (
  options: string[] | number[],
  defaultPageSize: string | number | undefined = undefined,
) => {
  if (defaultPageSize) {
    const optionAsInt = parseInt(defaultPageSize.toString());
    if (isNaN(optionAsInt)) return setFirstOption(options);
    return optionAsInt;
  } else {
    return setFirstOption(options);
  }
};

export function createPaginationStore<T>(
  pageSizeOptions: string[] | number[] = options,
  defaultPageSize: string | number | undefined = undefined,
): PaginationStore<T> {
  const initialPageSize = getInitialPageSize(pageSizeOptions, defaultPageSize);
  const paginationStore = writable({
    ...defaultStore,
    previousPageSize: initialPageSize,
    pageSize: initialPageSize,
  });
  const { set, update } = paginationStore;

  const pageSize = derived([page], ([$page]) => {
    const perPage = $page.url.searchParams.get(perPageKey);
    return perPage ? perPageFromSearchParameter(perPage) : undefined;
  });

  const { subscribe } = derived(
    [paginationStore, pageSize],
    ([$pagination, $pageSize]) => {
      const visibleItems =
        $pagination.indexData[$pagination.index]?.items ?? [];
      const indexStart = $pagination.indexData[$pagination.index]?.start ?? 0;
      const indexEnd = $pagination.indexData[$pagination.index]?.end ?? 0;
      const hasNextIndexData = Boolean(
        $pagination.indexData[$pagination.index + 1],
      );

      return {
        ...$pagination,
        visibleItems,
        indexStart,
        indexEnd,
        hasNextIndexData,
        pageSize: $pageSize ?? $pagination.pageSize,
      };
    },
  );

  const setNextPageWithItems = <T>(
    nextToken: string,
    items: T[],
    _store: PaginationItems<T>,
  ) => {
    const currentIndex = _store.index;
    const store = {
      ..._store,
      indexData: { ..._store.indexData },
      hasNext: Boolean(nextToken),
      updating: false,
      loading: false,
    };

    if (!items.length) return { ...store, hasNext: false };

    if (!store.indexData[currentIndex]) {
      store.indexData[currentIndex] = {
        nextToken,
        start: 1,
        end: items.length,
        items,
      };
      store.hasPrevious = false;
    } else {
      (store.index = currentIndex + 1),
        (store.indexData[store.index] = {
          nextToken,
          start: store.indexData[currentIndex].end + 1,
          end: store.indexData[currentIndex].end + items.length,
          items,
        });
      store.hasPrevious = true;
    }

    return store;
  };

  const setNextPage = <T>(_store: PaginationItems<T>) => {
    const store = {
      ..._store,
      index: _store.index + 1,
      hasPrevious: true,
      loading: false,
      updating: false,
    };

    if (!store.indexData[store.index]?.nextToken) {
      store.hasNext = false;
    }

    return store;
  };

  const setPreviousPage = <T>(_store: PaginationItems<T>) => {
    const store = {
      ..._store,
      hasNext: true,
      updating: false,
      loading: false,
      index: _store.index > 0 ? _store.index - 1 : 0,
    };

    if (store.index === 0) {
      store.hasPrevious = false;
    }

    return store;
  };

  const setNextRow = <T>(_store: PaginationItems<T>) => {
    const store = { ..._store };
    const indexLength = store.indexData[store.index]?.items?.length ?? 0;

    if (store.activeIndex < indexLength - 1) {
      store.activeIndex = store.activeIndex + 1;
    }

    return store;
  };

  const setPreviousRow = <T>(_store: PaginationItems<T>) => {
    const store = { ..._store };
    const activeIndex = store.activeIndex >= 1 ? store.activeIndex - 1 : 0;

    store.activeIndex = activeIndex;

    return store;
  };

  const setActiveIndex = <T>(
    _store: PaginationItems<T>,
    activeIndex: number,
  ) => {
    return { ..._store, activeIndex };
  };

  const resetPageSize = <T>(_store: PaginationItems<T>, pageSize: number) => {
    return {
      ..._store,
      pageSize,
      previousPageSize: pageSize,
      index: 0,
      indexData: {},
      loading: true,
      updating: false,
    };
  };

  return {
    subscribe,
    nextPageWithItems: (token: string, items: T[]) =>
      update((store) => setNextPageWithItems(token, items, store)),
    nextPage: () => update((store) => setNextPage(store)),
    previousPage: () => update((store) => setPreviousPage(store)),
    setUpdating: () => update((store) => ({ ...store, updating: true })),
    reset: () => set(defaultStore),
    resetPageSize: (pageSize) =>
      update((store) => resetPageSize(store, pageSize)),
    nextRow: () => update((store) => setNextRow(store)),
    previousRow: () => update((store) => setPreviousRow(store)),
    setActiveIndex: (activeIndex: number) =>
      update((store) => setActiveIndex(store, activeIndex)),
  };
}
