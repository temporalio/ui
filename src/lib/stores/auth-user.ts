import { get } from 'svelte/store';

import { persistStore } from '$lib/stores/persist-store';
import type { User } from '$lib/types/global';

export const authUser = persistStore<User>('AuthUser', {});

export const getAuthUser = (): User => get(authUser);

export const setAuthUser = (user: User) => {
  const { accessToken, idToken, name, email, picture } = user;

  if (!accessToken) {
    throw new Error('No access token');
  }

  authUser.set({
    accessToken,
    idToken,
    name,
    email,
    picture,
  });
};

export const clearAuthUser = () => {
  authUser.set({});
};
