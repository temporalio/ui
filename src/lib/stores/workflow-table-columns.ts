import { derived, type Readable } from 'svelte/store';

import { page } from '$app/stores';

import { namespaces } from './namespaces';
import { persistStore } from './persist-store';
import { customSearchAttributes } from './search-attributes';

export const MAX_PINNED_COLUMNS = 2;
export const WorkflowHeaderLabels = [
  'Status',
  'Workflow ID',
  'Run ID',
  'Type',
  'Start',
  'End',
  'History Size',
  'History Length',
  'Execution Time',
  'Execution Duration',
  'State Transitions',
  'Parent Namespace',
  'Parent Workflow ID',
  'Task Queue',
] as const;

export type WorkflowHeaderLabel = (typeof WorkflowHeaderLabels)[number];

// https://github.com/microsoft/TypeScript/issues/29729
// https://stackoverflow.com/a/61048124
// eslint-disable-next-line @typescript-eslint/ban-types
type AnyWorkflowHeaderLabel = WorkflowHeaderLabel | (string & {});

export type WorkflowHeader = {
  label: AnyWorkflowHeaderLabel;
  pinned: boolean;
};

type State = {
  [namespace: string]: WorkflowHeader[] | undefined;
};

type Action =
  | {
      type: 'WORKFLOW_COLUMN.ADD';
      payload: { label: AnyWorkflowHeaderLabel; namespace: string };
    }
  | {
      type: 'WORKFLOW_COLUMN.REMOVE';
      payload: { label: AnyWorkflowHeaderLabel; namespace: string };
    }
  | {
      type: 'WORKFLOW_COLUMN.PIN';
      payload: { label: AnyWorkflowHeaderLabel; namespace: string };
    }
  | {
      type: 'WORKFLOW_COLUMN.MOVE';
      payload: { from: number; to: number; namespace: string };
    };

const DEFAULT_COLUMNS: WorkflowHeader[] = [
  { label: 'Status', pinned: true },
  { label: 'Workflow ID', pinned: true },
  { label: 'Run ID', pinned: false },
  { label: 'Type', pinned: false },
  { label: 'Start', pinned: false },
  { label: 'End', pinned: false },
];

const DEFAULT_AVAILABLE_COLUMNS: WorkflowHeader[] = [
  { label: 'History Size', pinned: false },
  { label: 'History Length', pinned: false },
  { label: 'Execution Time', pinned: false },
  { label: 'Execution Duration', pinned: false },
  { label: 'State Transitions', pinned: false },
  { label: 'Parent Namespace', pinned: false },
  { label: 'Parent Workflow ID', pinned: false },
  { label: 'Task Queue', pinned: false },
];

export const getDefaultColumns = (): WorkflowHeader[] => {
  let columns: WorkflowHeader[];
  try {
    // try to get the list of columns that was stored last time they interacted
    // with the table before we made it namespace-specific
    const stringifiedOldColumns = window.localStorage.getItem(
      'workflow-table-columns',
    );
    const parsedOldColumns = JSON.parse(stringifiedOldColumns);

    if (stringifiedOldColumns && parsedOldColumns?.length) {
      columns = parsedOldColumns;
    } else {
      columns = DEFAULT_COLUMNS;
    }
  } catch {
    columns = DEFAULT_COLUMNS;
  }

  return columns;
};

export const persistedWorkflowTableColumns = persistStore<State>(
  'namespace-workflow-table-columns',
  {},
);

export const workflowTableColumns: Readable<State> = derived(
  [namespaces, page, persistedWorkflowTableColumns],
  ([$namespaces, $page, $persistedWorkflowTableColumns]) => {
    const state: State = {};

    const useOrAddDefaultTableColumnsToNamespace = (
      columns: State,
      namespace: string,
    ) => {
      if (!columns?.[namespace]?.length) {
        columns[namespace] = [...getDefaultColumns()];
        return columns[namespace];
      }
      return columns[namespace];
    };

    const namespaceColumns =
      $namespaces?.reduce(
        (namespaceToColumnsMap, { namespaceInfo: { name } }) => {
          return {
            ...namespaceToColumnsMap,
            [name]: useOrAddDefaultTableColumnsToNamespace(
              $persistedWorkflowTableColumns,
              name,
            ),
          };
        },
        state,
      ) ?? {};
    const { namespace: currentNamespace } = $page.params;

    return namespaceColumns[currentNamespace]
      ? namespaceColumns
      : {
          ...namespaceColumns,
          [currentNamespace]: useOrAddDefaultTableColumnsToNamespace(
            $persistedWorkflowTableColumns,
            currentNamespace,
          ),
        };
  },
);

export const pinnedColumnsWidth = persistStore<number>(
  'workflow-table-pinned-columns-width',
);

export const availableSystemSearchAttributeColumns: (
  namespace: string,
) => Readable<WorkflowHeader[]> = (namespace) =>
  derived(workflowTableColumns, ($workflowTableColumns) =>
    [...DEFAULT_COLUMNS, ...DEFAULT_AVAILABLE_COLUMNS].filter(
      (header) =>
        !$workflowTableColumns[namespace]?.some(
          (column) => column.label === header.label,
        ),
    ),
  );

export const availableCustomSearchAttributeColumns: (
  namespace: string,
) => Readable<WorkflowHeader[]> = (namespace: string) =>
  derived(
    [customSearchAttributes, workflowTableColumns],
    ([$customSearchAttributes, $workflowTableColumns]) =>
      Object.keys($customSearchAttributes)
        .filter(
          (searchAttribute) =>
            !$workflowTableColumns[namespace]?.some(
              (column) => column.label === searchAttribute,
            ),
        )
        .map((key) => ({
          label: key,
          pinned: false,
        })),
  );

const reducer = (action: Action, state: State): State => {
  switch (action.type) {
    case 'WORKFLOW_COLUMN.ADD': {
      const { label, namespace } = action.payload;
      const columns = state?.[namespace] ?? DEFAULT_COLUMNS;

      return {
        ...state,
        [namespace]: [...columns, { label, pinned: false }],
      };
    }
    case 'WORKFLOW_COLUMN.REMOVE': {
      const { label: labelToRemove, namespace } = action.payload;
      const columns = state?.[namespace] ?? DEFAULT_COLUMNS;

      return {
        ...state,
        [namespace]: columns.filter(({ label }) => label !== labelToRemove),
      };
    }
    case 'WORKFLOW_COLUMN.PIN': {
      const { label: labelToPin, namespace } = action.payload;
      const columns = state?.[namespace] ?? DEFAULT_COLUMNS;
      const index = columns.findIndex(({ label }) => label === labelToPin);

      const isPinned = columns[index].pinned;

      let lastPinned = -1;
      for (let i = columns.length - 1; i >= 0; i--) {
        if (columns[i].pinned) {
          lastPinned = i;
          break;
        }
      }

      const newColumns = [...columns];
      newColumns[index].pinned = !isPinned;

      if (index > lastPinned && !isPinned) {
        newColumns.splice(lastPinned + 1, 0, newColumns.splice(index, 1)[0]);
      } else if (index < lastPinned && isPinned) {
        newColumns.splice(lastPinned, 0, newColumns.splice(index, 1)[0]);
      }

      return {
        ...state,
        [namespace]: newColumns,
      };
    }
    case 'WORKFLOW_COLUMN.MOVE': {
      const { from, to, namespace } = action.payload;
      const columns = state?.[namespace] ?? DEFAULT_COLUMNS;
      const isPinned = columns[from].pinned;

      let lastPinned = 0;
      for (let i = columns.length - 1; i >= 0; i--) {
        if (columns[i].pinned) {
          lastPinned = i;
          break;
        }
      }

      const tempColumns = [...columns];
      if (to <= lastPinned && !isPinned) {
        tempColumns[from].pinned = true;
      } else if (to > lastPinned && isPinned) {
        tempColumns[from].pinned = false;
      }

      tempColumns.splice(to, 0, tempColumns.splice(from, 1)[0]);

      return {
        ...state,
        [namespace]: tempColumns.map((c, idx) =>
          idx > MAX_PINNED_COLUMNS - 1 ? { ...c, pinned: false } : c,
        ),
      };
    }
    default:
      return state;
  }
};

const dispatch = (action: Action) => {
  persistedWorkflowTableColumns.update((previousState) =>
    reducer(action, previousState),
  );
};

export const addColumn = (label: AnyWorkflowHeaderLabel, namespace: string) => {
  dispatch({ type: 'WORKFLOW_COLUMN.ADD', payload: { label, namespace } });
};

export const removeColumn = (
  label: AnyWorkflowHeaderLabel,
  namespace: string,
) => {
  dispatch({ type: 'WORKFLOW_COLUMN.REMOVE', payload: { label, namespace } });
};

export const moveColumn = (from: number, to: number, namespace: string) => {
  dispatch({ type: 'WORKFLOW_COLUMN.MOVE', payload: { from, to, namespace } });
};

export const pinColumn = (label: AnyWorkflowHeaderLabel, namespace: string) => {
  dispatch({ type: 'WORKFLOW_COLUMN.PIN', payload: { label, namespace } });
};
