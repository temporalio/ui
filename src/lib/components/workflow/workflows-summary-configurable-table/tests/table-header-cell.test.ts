import { afterEach, beforeEach, describe, expect, test } from 'vitest';

import { WorkflowHeaderLabels } from '$lib/stores/workflow-table-columns';

import TableHeaderCell from '../table-header-cell.svelte';

let target: HTMLElement;

beforeEach(() => {
  target = document.createElement('div');
  target.setAttribute('id', 'target');
  document.body.appendChild(target);
});

afterEach(() => {
  target.remove();
});

describe(TableHeaderCell.name, () => {
  test.each(WorkflowHeaderLabels)('%s renders', (label) => {
    const column = { label, pinned: false };
    const instance = new TableHeaderCell({ target, props: { column } });

    expect(instance).toBeTruthy();
    expect(target.innerHTML).toMatchSnapshot();
  });
});
