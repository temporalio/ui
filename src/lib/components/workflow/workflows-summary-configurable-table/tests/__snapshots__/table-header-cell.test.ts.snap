// Vitest Snapshot v1

exports[`Table_header_cell$ > End renders 1`] = `"<th class=\\"workflows-summary-table-header-cell svelte-1q2ffog\\" data-testid=\\"workflows-summary-table-header-cell-End\\">End</th>"`;

exports[`Table_header_cell$ > Execution Duration renders 1`] = `"<th class=\\"workflows-summary-table-header-cell svelte-1q2ffog\\" data-testid=\\"workflows-summary-table-header-cell-Execution Duration\\">Execution Duration</th>"`;

exports[`Table_header_cell$ > Execution Time renders 1`] = `"<th class=\\"workflows-summary-table-header-cell svelte-1q2ffog\\" data-testid=\\"workflows-summary-table-header-cell-Execution Time\\">Execution Time</th>"`;

exports[`Table_header_cell$ > History Length renders 1`] = `"<th class=\\"workflows-summary-table-header-cell svelte-1q2ffog\\" data-testid=\\"workflows-summary-table-header-cell-History Length\\">History Length</th>"`;

exports[`Table_header_cell$ > History Size renders 1`] = `"<th class=\\"workflows-summary-table-header-cell svelte-1q2ffog\\" data-testid=\\"workflows-summary-table-header-cell-History Size\\">History Size</th>"`;

exports[`Table_header_cell$ > Parent Namespace renders 1`] = `"<th class=\\"workflows-summary-table-header-cell svelte-1q2ffog\\" data-testid=\\"workflows-summary-table-header-cell-Parent Namespace\\">Parent Namespace</th>"`;

exports[`Table_header_cell$ > Parent Workflow ID renders 1`] = `"<th class=\\"workflows-summary-table-header-cell svelte-1q2ffog\\" data-testid=\\"workflows-summary-table-header-cell-Parent Workflow ID\\">Parent Workflow ID</th>"`;

exports[`Table_header_cell$ > Run ID renders 1`] = `"<th class=\\"workflows-summary-table-header-cell svelte-1q2ffog\\" data-testid=\\"workflows-summary-table-header-cell-Run ID\\"><div><div class=\\"relative inline\\" data-testid=\\"run-id-filter-button\\"><button type=\\"button\\" class=\\"icon-button svelte-1vs3on1\\" data-testid=\\"run-id-filter-button-button\\"><div class=\\"flex items-center gap-1 truncate svelte-1o0pqx5\\">Run ID <svg width=\\"20\\" height=\\"20\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"pointer-events-none\\"><title></title><path d=\\"M5.5 7.5H18.5C18.75 7.5 19 7.28125 19 7.03125C19 6.78125 18.75 6.5 18.5 6.5H5.5C5.21875 6.5 5 6.75 5 7C5 7.25 5.21875 7.5 5.5 7.5ZM16.5 11.5H7.5C7.21875 11.5 7 11.75 7 12C7 12.2812 7.21875 12.5 7.5 12.5H16.5C16.75 12.5 17 12.2812 17 12.0312C17 11.75 16.75 11.5 16.5 11.5ZM13.5 16.5H10.5C10.2188 16.5 10 16.75 10 17C10 17.25 10.2188 17.5 10.5 17.5H13.5C13.75 17.5 14 17.2812 14 17.0312C14 16.7812 13.75 16.5 13.5 16.5Z\\" fill=\\"currentcolor\\"></path></svg></div></button> </div> </div></th>"`;

exports[`Table_header_cell$ > Start renders 1`] = `"<th class=\\"workflows-summary-table-header-cell svelte-1q2ffog\\" data-testid=\\"workflows-summary-table-header-cell-Start\\">Start</th>"`;

exports[`Table_header_cell$ > State Transitions renders 1`] = `"<th class=\\"workflows-summary-table-header-cell svelte-1q2ffog\\" data-testid=\\"workflows-summary-table-header-cell-State Transitions\\">State Transitions</th>"`;

exports[`Table_header_cell$ > Status renders 1`] = `"<th class=\\"workflows-summary-table-header-cell svelte-1q2ffog\\" data-testid=\\"workflows-summary-table-header-cell-Status\\"><div><div class=\\"relative inline\\" data-testid=\\"execution-status-filter\\"><button type=\\"button\\" class=\\"icon-button svelte-1vs3on1\\" data-testid=\\"execution-status-filter-button\\"><div class=\\"flex items-center gap-1 truncate svelte-1o0pqx5\\">Status <svg width=\\"20\\" height=\\"20\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"pointer-events-none\\"><title></title><path d=\\"M5.5 7.5H18.5C18.75 7.5 19 7.28125 19 7.03125C19 6.78125 18.75 6.5 18.5 6.5H5.5C5.21875 6.5 5 6.75 5 7C5 7.25 5.21875 7.5 5.5 7.5ZM16.5 11.5H7.5C7.21875 11.5 7 11.75 7 12C7 12.2812 7.21875 12.5 7.5 12.5H16.5C16.75 12.5 17 12.2812 17 12.0312C17 11.75 16.75 11.5 16.5 11.5ZM13.5 16.5H10.5C10.2188 16.5 10 16.75 10 17C10 17.25 10.2188 17.5 10.5 17.5H13.5C13.75 17.5 14 17.2812 14 17.0312C14 16.7812 13.75 16.5 13.5 16.5Z\\" fill=\\"currentcolor\\"></path></svg></div></button> </div> </div></th>"`;

exports[`Table_header_cell$ > Task Queue renders 1`] = `"<th class=\\"workflows-summary-table-header-cell svelte-1q2ffog\\" data-testid=\\"workflows-summary-table-header-cell-Task Queue\\">Task Queue</th>"`;

exports[`Table_header_cell$ > Type renders 1`] = `"<th class=\\"workflows-summary-table-header-cell svelte-1q2ffog\\" data-testid=\\"workflows-summary-table-header-cell-Type\\"><div><div class=\\"relative inline\\" data-testid=\\"workflow-type-filter-button\\"><button type=\\"button\\" class=\\"icon-button svelte-1vs3on1\\" data-testid=\\"workflow-type-filter-button-button\\"><div class=\\"flex items-center gap-1 truncate svelte-1o0pqx5\\">Type <svg width=\\"20\\" height=\\"20\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"pointer-events-none\\"><title></title><path d=\\"M5.5 7.5H18.5C18.75 7.5 19 7.28125 19 7.03125C19 6.78125 18.75 6.5 18.5 6.5H5.5C5.21875 6.5 5 6.75 5 7C5 7.25 5.21875 7.5 5.5 7.5ZM16.5 11.5H7.5C7.21875 11.5 7 11.75 7 12C7 12.2812 7.21875 12.5 7.5 12.5H16.5C16.75 12.5 17 12.2812 17 12.0312C17 11.75 16.75 11.5 16.5 11.5ZM13.5 16.5H10.5C10.2188 16.5 10 16.75 10 17C10 17.25 10.2188 17.5 10.5 17.5H13.5C13.75 17.5 14 17.2812 14 17.0312C14 16.7812 13.75 16.5 13.5 16.5Z\\" fill=\\"currentcolor\\"></path></svg></div></button> </div> </div></th>"`;

exports[`Table_header_cell$ > Workflow ID renders 1`] = `"<th class=\\"workflows-summary-table-header-cell svelte-1q2ffog\\" data-testid=\\"workflows-summary-table-header-cell-Workflow ID\\"><div><div class=\\"relative inline\\" data-testid=\\"workflow-id-filter-button\\"><button type=\\"button\\" class=\\"icon-button svelte-1vs3on1\\" data-testid=\\"workflow-id-filter-button-button\\"><div class=\\"flex items-center gap-1 truncate svelte-1o0pqx5\\">Workflow ID <svg width=\\"20\\" height=\\"20\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"pointer-events-none\\"><title></title><path d=\\"M5.5 7.5H18.5C18.75 7.5 19 7.28125 19 7.03125C19 6.78125 18.75 6.5 18.5 6.5H5.5C5.21875 6.5 5 6.75 5 7C5 7.25 5.21875 7.5 5.5 7.5ZM16.5 11.5H7.5C7.21875 11.5 7 11.75 7 12C7 12.2812 7.21875 12.5 7.5 12.5H16.5C16.75 12.5 17 12.2812 17 12.0312C17 11.75 16.75 11.5 16.5 11.5ZM13.5 16.5H10.5C10.2188 16.5 10 16.75 10 17C10 17.25 10.2188 17.5 10.5 17.5H13.5C13.75 17.5 14 17.2812 14 17.0312C14 16.7812 13.75 16.5 13.5 16.5Z\\" fill=\\"currentcolor\\"></path></svg></div></button> </div> </div></th>"`;
